src/devices/bus/nscsi/tape.cpp

```diff
@@ -0,0 +1,1028 @@
+	save_item(NAME(m_r_suppress_bad_len));
+	save_item(NAME(m_rw_pending));
+	save_item(NAME(m_pl_buf));
+	save_item(NAME(m_pl_len));
+}
+
+void nscsi_tape_device::device_reset()
+{
+	nscsi_full_device::device_reset();
+	m_fixed_block_len = TAPE_BLOCK_LEN;
+	m_sequence_counter = m_image->sequence_counter();
+	m_tape_changed = false;
+	m_rw_pending = false;
+}
+
+void nscsi_tape_device::scsi_command()
+{
+	const int lun = get_lun(scsi_cmdbuf[1] >> 5); // LUN may be overridden by IDENTIFY, per SCSI-2 section 7.2.2
+	const int cmd = scsi_cmdbuf[0];
+	switch (cmd) {
+		case SC_INQUIRY:                        return handle_inquiry(lun);
+		case SC_REQUEST_SENSE:                  return handle_request_sense(lun);
+	}
+
+	if (m_sequence_counter != m_image->sequence_counter()) { // tape changed; we first report no tape even when there is new tape to let guest know old tape was removed
+		m_sequence_counter = m_image->sequence_counter();
+		m_tape_changed = true;
+		m_rw_pending = false;
+		return report_no_tape();
+	}
+
+	if (m_tape_changed) { // tape changed earlier
+		m_tape_changed = false;
+		return report_tape_changed();
+	}
+
+	if (lun) // error: we don't support LUNs other than 0 for other commands
+		return report_bad_lun(cmd, lun);
+
+	switch (cmd) {
+		case SC_MODE_SELECT_6:                  return handle_mode_select_6();
+		case SC_MODE_SENSE_6:                   return handle_mode_sense_6();
+		case SC_SEND_DIAGNOSTIC:                return handle_send_diagnostic();
+		case SC_TEST_UNIT_READY:                return handle_test_unit_ready();
+		case SC_PREVENT_ALLOW_MEDIUM_REMOVAL:   return handle_prevent_allow_medium_removal();
+		case SC_ERASE:                          return handle_erase();
+		case SC_LOAD_UNLOAD:                    return handle_load_unload();
+		case SC_LOCATE:                         return handle_locate();
+		case SC_READ_6:                         return handle_read_6();
+		case SC_READ_BLOCK_LIMITS:              return handle_read_block_limits();
+		case SC_READ_POSITION:                  return handle_read_position();
+		case SC_RELEASE_UNIT:                   return handle_release_unit();
+		case SC_RESERVE_UNIT:                   return handle_reserve_unit();
+		case SC_REWIND:                         return handle_rewind();
+		case SC_SPACE:                          return handle_space();
+		case SC_WRITE_6:                        return handle_write_6();
+		case SC_WRITE_FILEMARKS:                return handle_write_filemarks();
+		// TODO: support more optional commands, once we need to
+		case SC_CHANGE_DEFINITION:
+		case SC_COMPARE:
+		case SC_COPY:
+		case SC_COPY_AND_VERIFY:
+		case SC_LOG_SELECT:
+		case SC_LOG_SENSE:
+		case SC_MODE_SELECT_10:
+		case SC_MODE_SENSE_10:
+		case SC_READ_BUFFER:
+		case SC_READ_REVERSE:
+		case SC_RECEIVE_DIAGNOSTIC_RESULTS:
+		case SC_RECOVER_BUFFERED_DATA:
+		case SC_VERIFY_6:
+		case SC_WRITE_BUFFER:
+		default: // error: unsupported command
+			return report_bad_cmd(cmd);
```

Comment:
I think this should fall back to nscsi_full_device::scsi_command() after promoting set_sense_data() and helpers.

Reply:
I can do that, but there isn’t much point to it.  All `nscsi_full_device::scsi_command()` does is:
1. handle `SC_REZERO_UNIT`, which is a synonym for `SC_REWIND`, which we must handle on our own;
2. handle `SC_REQUEST_SENSE`, which we must handle on our own as well, due to the medium changing logic.